kratos/containers/constraint_data.h

```
@@ -0,0 +1,407 @@
+    double ConstantUpdate() const {return constantUpdate;}
+    unsigned int SlaveDofId() const { return slaveDofId; }
+    unsigned int SlaveDofKey() const { return slaveDofKey; }
+    unsigned int SlaveEquationId() const { return slaveEquationId; }
+
+    // Add a master to this slave given are the masterDofId, masterDofKey, weight
+    void AddMasterData(unsigned int masterDofId, std::size_t masterDofKey, double weight)
+    {
+        auto it = std::find_if(mMasterData.begin(), mMasterData.end(), [&masterDofId](const MasterData &obj) { return obj.masterDofId == masterDofId; });
+        if (it == mMasterData.end()) // No entry for this master dof ID. So we add the data
+        {
+            mMasterData.push_back(MasterData(masterDofId, masterDofKey, weight));
+        }
+        else
+        { // The dof is perviously added(probably multiple times), but not sure if with the same key.
+            auto it = mMasterData.begin();
+            auto condLambda = [&masterDofId](const MasterData &obj) { return obj.masterDofId == masterDofId; };
+            while ((it != mMasterData.end())) // We check the key for all the occurances of this dofId
+            {
+                if (it->masterKey == masterDofKey) // Same pair of dofId and dofKey already exists. So add the weights.
+                {
+                    it->masterWeight += weight;
+                    break;
+                }
+                it = std::find_if(std::next(it), mMasterData.end(), condLambda);
+            }
+
+            if (it == mMasterData.end()) // Here the dofId is added but not with the given key. So we add the data to the vectors.
+            {
+                mMasterData.push_back(MasterData(masterDofId, masterDofKey, weight));
+            }
+        }
+    }
+
+    // For this slave, add a masterEquationId to a master given by masterDofID and masterDofKey
+    void SetMasterEquationId(unsigned int masterDofId, std::size_t masterDofKey, unsigned int masterEquationId)

```

Same here. If you had an additional parent class that only contains DofId and DofKey, then you could define a `==` for it and use this here or you would need to define a compare operator for the ConstraintData that ignores the EquationId. Also you might want to throw an exception here if the search fails.